tests/e2e/pubsub/pubsub_test.go

```diff
@@ -0,0 +1,189 @@
+// ------------------------------------------------------------
+
+package pubsubapp_e2e
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"log"
+	"math/rand"
+	"net/http"
+	"os"
+	"reflect"
+	"testing"
+
+	
+	"github.com/dapr/dapr/tests/e2e/utils"
+	kube "github.com/dapr/dapr/tests/platforms/kubernetes"
+	"github.com/dapr/dapr/tests/runner"
+	"github.com/stretchr/testify/require"
+)
+
+const numHealthChecks = 3      // Number of get calls before starting tests.
+const testManyEntriesCount = 5 // Anything between 1 and the number above (inclusive).
+
+var tr *runner.TestRunner
+
+const (
+	publisherAppName  = "pubsub-publisher"
+	subscriberAppName = "pubsub-subscriber"
+)
+
+type testCommandRequest struct {
+	Message string `json:"message,omitempty"`
+}
+
+// sent to the publisher app, which will publish data to dapr
+type publishCommand struct {
+	Topic string `json:"topic"`
+	Data  string `json:"data"`
+}
+
+type appResponse struct {
+	Message   string `json:"message,omitempty"`
+	StartTime int    `json:"start_time,omitempty"`
+	EndTime   int    `json:"end_time,omitempty"`
+}
+
+// data returned from the subscriber app
+type receivedMessagesResponse struct {
+	ReceivedByTopicA []string `json:"pubsub-a-topic"`
+	ReceivedByTopicB []string `json:"pubsub-b-topic"`
+}
+
+var receivedMessages []string
+
+// indexHandler is the handler for root path
+func indexHandler(w http.ResponseWriter, r *http.Request) {
+	log.Printf("indexHandler is called\n")
+
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(appResponse{Message: "OK"})
+}
+
+// sends messages to the publisher app.  The publisher app does the actual publish
+func sendToPublisher(t *testing.T, publisherExternalURL string, topic string) ([]string, error) {
+	var sentMessages []string
+	commandBody := publishCommand{Topic: topic}
+	offset := rand.Intn(100)
+	for i := offset; i < offset+100; i++ {
+		commandBody.Data = fmt.Sprintf("message-%d", i)
+
+		sentMessages = append(sentMessages, commandBody.Data)
+		jsonValue, err := json.Marshal(commandBody)
+		require.NoError(t, err)
+	
+		// this is the publish app's endpoint, not a dapr endpoint
+		url := fmt.Sprintf("http://%s/tests/publish", publisherExternalURL)
+
+		// debuggability - trace info about the first message.  don't trace others so it doesn't flood log.
+		if i == offset {
+			log.Printf("Sending first publish app at url %s and body '%s', this log will not print for subsequent messages for same topic", url, jsonValue)
+		}
+
+		postResp, err := http.Post(url, "application/json", bytes.NewBuffer(jsonValue))
+		require.NoError(t, err)
+		if err != nil {
+			if postResp != nil {
+				log.Printf("Publish failed with error=%s, StatusCode=%d", err.Error(), postResp.StatusCode)
+			} else {
+				log.Printf("Publish failed with error=%s, response is nil", err.Error())
+			}
+
+			return nil, err
+		}
+	}
+
+	return sentMessages, nil
+}
+
+func doIt(t *testing.T, publisherExternalURL string) receivedMessagesResponse {
```

Comment:
doIt does not really mean much here. Maybe publishAll() or something.

Reply:
Yeah, meant to change it before sending PR...modified to be more descriptive.